implemented 'update costmap on request' option

JohnTGZ opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

This PR targets #3121, adding the 'update costmap on request' option via the parameter `update_on_request` which can be specified on the costmap layer, as shown in [nav2_bringup/params/nav2_params.yaml](https://github.com/JohnTGZ/navigation2/blob/0917791a0be38e52241acafa93b13459b726713e/nav2_bringup/params/nav2_params.yaml#L185).

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (#3121) |
| Primary OS tested on | (Ubuntu 22.04) |
| Robotic platform tested on | (turtlebot gazebo simulation) |

---

## Description of contribution in a few bullet points

- The implementation is such that if `update_on_request` is set to `TRUE`, the value of `update_frequency` is set to `0.0`, therefore disabling the costmap update loop on the thread. 
- Then, planner server will update the costmap on each call to [`waitForCostmap()`](https://github.com/JohnTGZ/navigation2/blob/0917791a0be38e52241acafa93b13459b726713e/nav2_planner/src/planner_server.cpp#L251-L263) instead of the costmap being updated on a fixed loop. 
- Similarly, the costmap for the controller server will be updated the call to [`computeControl()`](https://github.com/JohnTGZ/navigation2/blob/0917791a0be38e52241acafa93b13459b726713e/nav2_controller/src/controller_server.cpp#L387-390).
- If `update_on_request` is `FALSE`, the behaviour is as default.

<!--
* I added this neat new feature
* Also fixed a typo in a parameter name in nav2_costmap_2d
-->

## Description of documentation updates required from your changes

<!--
* Added new parameter, so need to add that to default configs and documentation page
* I added some capabilities, need to document them
-->

Will need to update the documentation for the `update_on_request` parameter and how setting it to `TRUE` will affect the value of `update_frequency` internally.

---

## Future work that may be required in bullet points

If `update_on_request` is set to `TRUE`, it is possible to disable initializing the thread for the map update loop as it won't be utilised.

<!--
* I think there might be some optimizations to be made from STL vector
* I see alot of redundancy in this package, we might want to add a function `bool XYZ()` to reduce clutter
* I tested on a differential drive robot, but there might be issues turning near corners on an omnidirectional platform
-->

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

AlexeyMerzlyakov commented on nav2_planner/src/planner_server.cpp:
```diff
@@ -250,6 +250,11 @@ bool PlannerServer::isServerInactive(
+  if (costmap_ros_->isUpdateOnRequest()){
+    RCLCPP_DEBUG(get_logger(), "Updating costmap on request!");
+    costmap_ros_->updateMapOnRequest();
+    return;
```
Why return should be here?

JohnTGZ replied:
My mistake. The return should not be there because the function should still continue to wait until the costmap has been fully updated

AlexeyMerzlyakov commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -384,10 +384,16 @@ void ControllerServer::computeControl()
-      rclcpp::Rate r(100);
-      while (!costmap_ros_->isCurrent()) {
-        r.sleep();
+      if (costmap_ros_->isUpdateOnRequest()){
+        RCLCPP_DEBUG(get_logger(), "Updating costmap on request!");
+        costmap_ros_->updateMapOnRequest();
+      }
+      else {
+        // Don't compute a trajectory until costmap is valid (after clear costmap)
+        rclcpp::Rate r(100);
+        while (!costmap_ros_->isCurrent()) {
+          r.sleep();
+        }
```
We need to check if we are updating on request (that is manually), mean we have no updating loop thread enabled. Thus, in this case I believe we need to 1. update map manually by calling new API and then 2. wait until costmap is being updated. But I see these two steps are in separate if-branches. Why this logic was introduced here?

AlexeyMerzlyakov reviewed (commented):
@JohnTGZ, Thank you for taking care on this issue. Please refer to the review items. And then make a PR to a Nav2 `main` branch instead of `humble` (since we are working with `main` branch in Nav2 first).

JohnTGZ commented on nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp:
```diff
@@ -302,6 +302,21 @@ class Costmap2DROS : public nav2_util::LifecycleNode
+   * @brief Update the and publish the costmap once.
+   * 
+   */
+  void updateMapOnRequest();
```
I was thinking something along the lines of `updateAndPublishMap()`, would that convey the meaning better?

AlexeyMerzlyakov replied:
Yes, I believe `updateAndPublishMap()` seems to have representing its actual and more general state w/o binding to any specific use-case

JohnTGZ commented on nav2_costmap_2d/src/costmap_2d_ros.cpp:
```diff
@@ -323,6 +327,7 @@ Costmap2DROS::getParameters()
   get_parameter("width", map_width_meters_);
   get_parameter("plugins", plugin_names_);
   get_parameter("filters", filter_names_);
+  get_parameter("update_on_request", update_on_request_);
```
I think that means we will probably have to handle starting/stopping the mapUpdateLoop() thread too.

AlexeyMerzlyakov replied:
The handling logic for this parameter dynamic parameters callback was initially missed. At this logic, yes, you can create or join the threat running `mapUpdateLoop()` depending on `update_on_request` state.

JohnTGZ commented on nav2_costmap_2d/src/costmap_2d_ros.cpp:
```diff
@@ -242,6 +243,9 @@ Costmap2DROS::on_activate(const rclcpp_lifecycle::State & /*state*/)
   map_update_thread_shutdown_ = false;
+
+  updateMapOnRequest();
+  map_update_frequency_ = update_on_request_ ? 0.0 : map_update_frequency_;
```
That makes sense, I must have missed it. Thanks for highlighting this

JohnTGZ replied:
Also, are you able to clarify what you mean by moving the thread creation/joining into a separate sub-routine? Any example would be helpful

AlexeyMerzlyakov replied:
I mean some simple routine, with code like:
```
Costmap2DROS::mapUpdateThreadOn() {
  map_update_thread_shutdown_ = false;
  map_update_thread_ = std::make_unique<std::thread>(
    std::bind(&Costmap2DROS::mapUpdateLoop, this, map_update_frequency_));
}

Costmap2DROS::mapUpdateThreadOff() {
  map_update_thread_shutdown_ = true;
  map_update_thread_->join();

}
```
Or whenever you want.
Since these are only two lines, it is not mandatory: you could add these lines just directly in all required places (`on_activate()`, `on_deactivate()` and `dynamicParametersCallback()`).

JohnTGZ replied:
Okay I see what you mean. Yeah I think a routine would make things much clearer, I will incorporate this

JohnTGZ commented on nav2_costmap_2d/src/costmap_2d_ros.cpp:
```diff
@@ -242,6 +243,9 @@ Costmap2DROS::on_activate(const rclcpp_lifecycle::State & /*state*/)
   stop_updates_ = false;
   map_update_thread_shutdown_ = false;
+
+  updateMapOnRequest();
```
Thanks for pointing this out, I studied the code in more detail and this is an important detail to take note of

JohnTGZ commented:
@AlexeyMerzlyakov Thanks for the feedback, I have studied the code in more detail to understand the implications behind the code Im changing. I have addressed these comments in another branch that merges with `main`. Should I close this branch and open another PR?

AlexeyMerzlyakov commented:
> Should I close this branch and open another PR?
Once you will resolve all review items from this PR, please open a new one for `main` branch.

JohnTGZ commented:
This PR has been shifted to #3300 as this PR is for the wrong branch.
